Panic creating ConfigMap.

Please forgive the convoluted example but I'm getting a panic from the following code. I've stripped the functionality down to the bare minimum to cause it. If I create the KeyVault in one update and then create the configmap in the second it all works.

The method to get the kube config from a keyvault works everywhere else, but fails in this case. 

Panic output is at the bottom.


```typescript
import * as pulumi from "@pulumi/pulumi";
import * as azure from "@pulumi/azure"
import * as kubernetesInput from "@pulumi/kubernetes/types/input";
import * as kubernetes from "@pulumi/kubernetes";

export const main = async () => {

    const testKeyVault = new azure.keyvault.KeyVault("kv-test", {
        name: "kv-test",
        resourceGroupName: "Infrastructure",
        location: "westeurope",
        sku: { name: "standard" },
        tenantId: azure.config.tenantId || ""
    });

    const getConfigMapSpec = (
        namespace: pulumi.Input<string>,
        key: pulumi.Input<string>,
        contents: pulumi.Input<Object>,
        labels: any = {},
    ): kubernetesInput.core.v1.ConfigMap => pulumi.all([
        key,
        pulumi.output(contents).apply(x => JSON.stringify(x))
    ]).apply(([
        fileName,
        contents
    ]) => {
        let data: { [key: string]: string; } = {};
        data[fileName] = contents;

        return ({
            metadata: {
                namespace: namespace,
                labels: labels,
            },
            data: {
                ...data
            }
        });
    });

    const config = pulumi.all([
        testKeyVault.vaultUri
    ]).apply(([
        keyVaultUri
    ]) =>
        new Object({
            KeyVault: { BaseUrl: keyVaultUri }
        })
    );

    const getProvider = async (
        keyVault: azure.keyvault.GetKeyVaultResult,
        kubeConfigSecretName: string
    ): Promise<kubernetes.Provider> => new Promise(async (resolve, reject) => {
        const kubeConfig = pulumi.output(keyVault.id).apply(x => azure.keyvault.getSecret({
            keyVaultId: x,
            name: kubeConfigSecretName
        }));
        pulumi.output(kubeConfig.value).apply(x => {
            console.log(x);
            resolve(new kubernetes.Provider("Cluster", { kubeconfig: x }))
        });
    });

    const coreKeyVault = await azure.keyvault.getKeyVault({ name: "test", resourceGroupName: "test" });
    const provider = await getProvider(coreKeyVault, "Cluster-KubeConfig");

    new kubernetes.core.v1.ConfigMap(
        "test-configmap",
        getConfigMapSpec("default", "appsettings.json", config),
        { provider: provider }
    );

};
main();

```

```
panic: interface conversion: interface {} is resource.Computed, not map[string]interface {}
    goroutine 159 [running]:
    github.com/pulumi/pulumi-kubernetes/pkg/metadata.SetAnnotation(...)
        /home/travis/gopath/src/github.com/pulumi/pulumi-kubernetes/pkg/metadata/annotations.go:52
    github.com/pulumi/pulumi-kubernetes/pkg/metadata.SetAnnotationTrue(...)
        /home/travis/gopath/src/github.com/pulumi/pulumi-kubernetes/pkg/metadata/annotations.go:67
    github.com/pulumi/pulumi-kubernetes/pkg/metadata.AssignNameIfAutonamable(0xc00087f978, 0xc0010e80a7, 0x14)
        /home/travis/gopath/src/github.com/pulumi/pulumi-kubernetes/pkg/metadata/naming.go:35 +0x385
    github.com/pulumi/pulumi-kubernetes/pkg/provider.(*kubeProvider).Check(0xc0000e9e00, 0x1950600, 0xc0010ea090, 0xc00091af40, 0xc0000e9e00, 0x1523301, 0xc00091af80)
        /home/travis/gopath/src/github.com/pulumi/pulumi-kubernetes/pkg/provider/provider.go:269 +0x15aa
    github.com/pulumi/pulumi-kubernetes/vendor/github.com/pulumi/pulumi/sdk/proto/go._ResourceProvider_Check_Handler.func1(0x1950600, 0xc0010ea090, 0x1652c80, 0xc00091af40, 0x1671ea0, 0x278d518, 0x1950600, 0xc0010ea090)
        /home/travis/gopath/src/github.com/pulumi/pulumi-kubernetes/vendor/github.com/pulumi/pulumi/sdk/proto/go/provider.pb.go:1303 +0x8d
    github.com/pulumi/pulumi-kubernetes/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1(0x1950600, 0xc0010b5ec0, 0x1652c80, 0xc00091af40, 0xc00093eb40, 0xc00093eb60, 0x0, 0x0, 0x19136e0, 0xc0003f04d0)
        /home/travis/gopath/src/github.com/pulumi/pulumi-kubernetes/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc/server.go:61 +0x367
    github.com/pulumi/pulumi-kubernetes/vendor/github.com/pulumi/pulumi/sdk/proto/go._ResourceProvider_Check_Handler(0x1673660, 0xc0000e9e00, 0x1950600, 0xc0010b5ec0, 0xc0009405f0, 0xc0003f2ea0, 0x1950600, 0xc0010b5ec0, 0xc000a5c780, 0x125)
        /home/travis/gopath/src/github.com/pulumi/pulumi-kubernetes/vendor/github.com/pulumi/pulumi/sdk/proto/go/provider.pb.go:1305 +0x15f
    github.com/pulumi/pulumi-kubernetes/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc000461200, 0x196b800, 0xc000437500, 0xc00050b100, 0xc0003fd3b0, 0x275fa40, 0x0, 0x0, 0x0)
        /home/travis/gopath/src/github.com/pulumi/pulumi-kubernetes/vendor/google.golang.org/grpc/server.go:972 +0x477
    github.com/pulumi/pulumi-kubernetes/vendor/google.golang.org/grpc.(*Server).handleStream(0xc000461200, 0x196b800, 0xc000437500, 0xc00050b100, 0x0)
        /home/travis/gopath/src/github.com/pulumi/pulumi-kubernetes/vendor/google.golang.org/grpc/server.go:1252 +0xdad
    github.com/pulumi/pulumi-kubernetes/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0003a4010, 0xc000461200, 0x196b800, 0xc000437500, 0xc00050b100)
        /home/travis/gopath/src/github.com/pulumi/pulumi-kubernetes/vendor/google.golang.org/grpc/server.go:691 +0xa6
    created by github.com/pulumi/pulumi-kubernetes/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
        /home/travis/gopath/src/github.com/pulumi/pulumi-kubernetes/vendor/google.golang.org/grpc/server.go:689 +0xa8
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,11 @@
 ## 0.23.2 (Unreleased)
 
+### Improvements
+
+- Fix an issue where the provider would panic during preview if and object's
+  `metadata` property ended up being computed from one or more unknown
+  values. (Fixes [#559](https://github.com/pulumi/pulumi-kubernetes/issues/559)).
+
 ## 0.23.1 (May 10, 2019)
 
 ### Supported Kubernetes versions

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,10 +1,16 @@
 ## 0.23.2 (Unreleased)
 
+### Major changes
+
+-   None
+
 ### Improvements
 
-- Fix an issue where the provider would panic during preview if and object's
-  `metadata` property ended up being computed from one or more unknown
-  values. (Fixes [#559](https://github.com/pulumi/pulumi-kubernetes/issues/559)).
+-   None
+
+### Bug fixes
+
+-   Fix panics during preview when `metadata` is a computed value (https://github.com/pulumi/pulumi-kubernetes/pull/572)
 
 ## 0.23.1 (May 10, 2019)
 

--- a/pkg/metadata/annotations.go
+++ b/pkg/metadata/annotations.go
@@ -17,6 +17,7 @@ package metadata
 import (
 	"strings"
 
+	"github.com/pulumi/pulumi/pkg/resource"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
@@ -48,9 +49,17 @@ func IsInternalAnnotation(key string) bool {
 // SetAnnotation sets the specified key, value annotation on the provided Unstructured object.
 func SetAnnotation(obj *unstructured.Unstructured, key, value string) {
 	// Note: Cannot use obj.GetAnnotations() here because it doesn't properly handle computed values from preview.
+	// during preview, our strategy for if metdata or annotations end up being computed values, is to just not
+	// apply an annotiation (since there's no way to insert data into the computed object)
 	metadataRaw := obj.Object["metadata"]
+	if isComputedValue(metadataRaw) {
+		return
+	}
 	metadata := metadataRaw.(map[string]interface{})
 	annotationsRaw, ok := metadata["annotations"]
+	if isComputedValue(annotationsRaw) {
+		return
+	}
 	var annotations map[string]interface{}
 	if !ok {
 		annotations = make(map[string]interface{})
@@ -78,4 +87,9 @@ func IsAnnotationTrue(obj *unstructured.Unstructured, key string) bool {
 func GetAnnotationValue(obj *unstructured.Unstructured, key string) string {
 	annotations := obj.GetAnnotations()
 	return annotations[key]
+}
+
+func isComputedValue(v interface{}) bool {
+	_, isComputed := v.(resource.Computed)
+	return isComputed
 }

--- a/pkg/metadata/annotations.go
+++ b/pkg/metadata/annotations.go
@@ -49,8 +49,8 @@ func IsInternalAnnotation(key string) bool {
 // SetAnnotation sets the specified key, value annotation on the provided Unstructured object.
 func SetAnnotation(obj *unstructured.Unstructured, key, value string) {
 	// Note: Cannot use obj.GetAnnotations() here because it doesn't properly handle computed values from preview.
-	// during preview, our strategy for if metdata or annotations end up being computed values, is to just not
-	// apply an annotiation (since there's no way to insert data into the computed object)
+	// During preview, don't set annotations if the metadata or annotation contains a computed value since there's
+	// no way to insert data into the computed object.
 	metadataRaw := obj.Object["metadata"]
 	if isComputedValue(metadataRaw) {
 		return

--- a/pkg/metadata/annotations_test.go
+++ b/pkg/metadata/annotations_test.go
@@ -1,0 +1,45 @@
+// Copyright 2016-2019, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metadata
+
+import (
+	"testing"
+
+	"github.com/pulumi/pulumi/pkg/resource"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func TestSetAnnotationMetadataComputed(t *testing.T) {
+	obj := &unstructured.Unstructured{Object: map[string]interface{}{
+		"metadata": resource.Computed{Element: resource.NewObjectProperty(nil)},
+	}}
+
+	// Since metadata is a computed property, we can't really set an annotation of the object, but we should not fail
+	// as the metadata property of an object could be computed during previews.
+	SetAnnotation(obj, "foo", "bar")
+}
+
+func TestSetAnnotationMetadataAnnotationsComputed(t *testing.T) {
+	obj := &unstructured.Unstructured{Object: map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"annotations": resource.Computed{Element: resource.NewObjectProperty(nil)},
+		},
+	}}
+
+	// Since metadata is a computed property, we can't really set an annotation of the object, but we should not fail
+	// as the metadata property of an object could be computed during previews.
+	SetAnnotation(obj, "foo", "bar")
+}

--- a/pkg/metadata/annotations_test.go
+++ b/pkg/metadata/annotations_test.go
@@ -18,28 +18,62 @@ import (
 	"testing"
 
 	"github.com/pulumi/pulumi/pkg/resource"
-
+	"github.com/stretchr/testify/assert"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
-func TestSetAnnotationMetadataComputed(t *testing.T) {
-	obj := &unstructured.Unstructured{Object: map[string]interface{}{
+func TestSetAnnotation(t *testing.T) {
+	noAnnotationObj := &unstructured.Unstructured{Object: map[string]interface{}{
+		"metadata": map[string]interface{}{},
+	}}
+	existingAnnotationObj := &unstructured.Unstructured{Object: map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"annotations": map[string]interface{}{
+				"pulumi": "rocks",
+			},
+		},
+	}}
+	computedMetadataObj := &unstructured.Unstructured{Object: map[string]interface{}{
 		"metadata": resource.Computed{Element: resource.NewObjectProperty(nil)},
 	}}
-
-	// Since metadata is a computed property, we can't really set an annotation of the object, but we should not fail
-	// as the metadata property of an object could be computed during previews.
-	SetAnnotation(obj, "foo", "bar")
-}
-
-func TestSetAnnotationMetadataAnnotationsComputed(t *testing.T) {
-	obj := &unstructured.Unstructured{Object: map[string]interface{}{
+	computedAnnotationObj := &unstructured.Unstructured{Object: map[string]interface{}{
 		"metadata": map[string]interface{}{
 			"annotations": resource.Computed{Element: resource.NewObjectProperty(nil)},
 		},
 	}}
 
-	// Since metadata is a computed property, we can't really set an annotation of the object, but we should not fail
-	// as the metadata property of an object could be computed during previews.
-	SetAnnotation(obj, "foo", "bar")
+	type args struct {
+		obj         *unstructured.Unstructured
+		key         string
+		value       string
+		expectSet   bool // True if SetAnnotation is expected to set the annotation.
+		expectKey   string
+		expectValue string
+	}
+	tests := []struct {
+		name string
+		args args
+	}{
+		{"set-with-no-annotation", args{
+			obj: noAnnotationObj, key: "foo", value: "bar", expectSet: true, expectKey: "foo", expectValue: "bar"}},
+		{"set-with-existing-annotations", args{
+			obj: existingAnnotationObj, key: "foo", value: "bar", expectSet: true, expectKey: "foo", expectValue: "bar"}},
+
+		// Computed fields cannot be set, so SetAnnotation is a no-op.
+		{"set-with-computed-metadata", args{
+			obj: computedMetadataObj, key: "foo", value: "bar", expectSet: false}},
+		{"set-with-computed-annotation", args{
+			obj: computedAnnotationObj, key: "foo", value: "bar", expectSet: false}},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			SetAnnotation(tt.args.obj, tt.args.key, tt.args.value)
+			annotations := tt.args.obj.GetAnnotations()
+			value, ok := annotations[tt.args.expectKey]
+			assert.Equal(t, tt.args.expectSet, ok)
+			if ok {
+				assert.Equal(t, tt.args.expectValue, value)
+			}
+		})
+	}
 }

--- a/pkg/metadata/labels.go
+++ b/pkg/metadata/labels.go
@@ -21,8 +21,8 @@ import (
 // SetLabel sets the specified key/value pair as a label on the provided Unstructured object.
 func SetLabel(obj *unstructured.Unstructured, key, value string) {
 	// Note: Cannot use obj.GetLabels() here because it doesn't properly handle computed values from preview.
-	// during preview, our strategy for if metdata or annotations end up being computed values, is to just not
-	// apply an annotiation (since there's no way to insert data into the computed object)
+	// During preview, don't set labels if the metadata or label contains a computed value since there's
+	// no way to insert data into the computed object.
 	metadataRaw := obj.Object["metadata"]
 	if isComputedValue(metadataRaw) {
 		return

--- a/pkg/metadata/labels.go
+++ b/pkg/metadata/labels.go
@@ -21,9 +21,17 @@ import (
 // SetLabel sets the specified key/value pair as a label on the provided Unstructured object.
 func SetLabel(obj *unstructured.Unstructured, key, value string) {
 	// Note: Cannot use obj.GetLabels() here because it doesn't properly handle computed values from preview.
+	// during preview, our strategy for if metdata or annotations end up being computed values, is to just not
+	// apply an annotiation (since there's no way to insert data into the computed object)
 	metadataRaw := obj.Object["metadata"]
+	if isComputedValue(metadataRaw) {
+		return
+	}
 	metadata := metadataRaw.(map[string]interface{})
 	labelsRaw, ok := metadata["labels"]
+	if isComputedValue(labelsRaw) {
+		return
+	}
 	var labels map[string]interface{}
 	if !ok {
 		labels = make(map[string]interface{})

--- a/pkg/metadata/labels_test.go
+++ b/pkg/metadata/labels_test.go
@@ -1,0 +1,45 @@
+// Copyright 2016-2019, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metadata
+
+import (
+	"testing"
+
+	"github.com/pulumi/pulumi/pkg/resource"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func TestSetLabelMetadataComputed(t *testing.T) {
+	obj := &unstructured.Unstructured{Object: map[string]interface{}{
+		"metadata": resource.Computed{Element: resource.NewObjectProperty(nil)},
+	}}
+
+	// Since metadata is a computed property, we can't really set an annotation of the object, but we should not fail
+	// as the metadata property of an object could be computed during previews.
+	SetLabel(obj, "foo", "bar")
+}
+
+func TestSetLabelMetadataLabelsComputed(t *testing.T) {
+	obj := &unstructured.Unstructured{Object: map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"labels": resource.Computed{Element: resource.NewObjectProperty(nil)},
+		},
+	}}
+
+	// Since metadata is a computed property, we can't really set an annotation of the object, but we should not fail
+	// as the metadata property of an object could be computed during previews.
+	SetLabel(obj, "foo", "bar")
+}

--- a/pkg/metadata/labels_test.go
+++ b/pkg/metadata/labels_test.go
@@ -18,28 +18,62 @@ import (
 	"testing"
 
 	"github.com/pulumi/pulumi/pkg/resource"
-
+	"github.com/stretchr/testify/assert"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
-func TestSetLabelMetadataComputed(t *testing.T) {
-	obj := &unstructured.Unstructured{Object: map[string]interface{}{
+func TestSetLabel(t *testing.T) {
+	noLabelObj := &unstructured.Unstructured{Object: map[string]interface{}{
+		"metadata": map[string]interface{}{},
+	}}
+	existingLabelObj := &unstructured.Unstructured{Object: map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"labels": map[string]interface{}{
+				"pulumi": "rocks",
+			},
+		},
+	}}
+	computedMetadataObj := &unstructured.Unstructured{Object: map[string]interface{}{
 		"metadata": resource.Computed{Element: resource.NewObjectProperty(nil)},
 	}}
-
-	// Since metadata is a computed property, we can't really set an annotation of the object, but we should not fail
-	// as the metadata property of an object could be computed during previews.
-	SetLabel(obj, "foo", "bar")
-}
-
-func TestSetLabelMetadataLabelsComputed(t *testing.T) {
-	obj := &unstructured.Unstructured{Object: map[string]interface{}{
+	computedLabelObj := &unstructured.Unstructured{Object: map[string]interface{}{
 		"metadata": map[string]interface{}{
 			"labels": resource.Computed{Element: resource.NewObjectProperty(nil)},
 		},
 	}}
 
-	// Since metadata is a computed property, we can't really set an annotation of the object, but we should not fail
-	// as the metadata property of an object could be computed during previews.
-	SetLabel(obj, "foo", "bar")
+	type args struct {
+		obj         *unstructured.Unstructured
+		key         string
+		value       string
+		expectSet   bool // True if SetLabel is expected to set the label.
+		expectKey   string
+		expectValue string
+	}
+	tests := []struct {
+		name string
+		args args
+	}{
+		{"set-with-no-label", args{
+			obj: noLabelObj, key: "foo", value: "bar", expectSet: true, expectKey: "foo", expectValue: "bar"}},
+		{"set-with-existing-labels", args{
+			obj: existingLabelObj, key: "foo", value: "bar", expectSet: true, expectKey: "foo", expectValue: "bar"}},
+
+		// Computed fields cannot be set, so SetLabel is a no-op.
+		{"set-with-computed-metadata", args{
+			obj: computedMetadataObj, key: "foo", value: "bar", expectSet: false}},
+		{"set-with-computed-label", args{
+			obj: computedLabelObj, key: "foo", value: "bar", expectSet: false}},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			SetLabel(tt.args.obj, tt.args.key, tt.args.value)
+			labels := tt.args.obj.GetLabels()
+			value, ok := labels[tt.args.expectKey]
+			assert.Equal(t, tt.args.expectSet, ok)
+			if ok {
+				assert.Equal(t, tt.args.expectValue, value)
+			}
+		})
+	}
 }

